backend/src/controllers/v1/secretsFolderController.ts

```diff
@@ -1,107 +1,233 @@
+      },
+    });
+    await folder.save();
+    const folderVersion = new FolderVersion({
+      workspace: workspaceId,
+      environment,
+      nodes: folder.nodes,
+    });
+    await folderVersion.save();
+    return res.json({ folder: { id, name: folderName } });
   }
 
-  const currentFolderPath = completePath + "/" + folderName // construct new path with current folder to be created
-  const normalizedCurrentPath = normalizePath(currentFolderPath)
-  const normalizedParentPath = getParentPath(normalizedCurrentPath)
+  const folder = appendFolder(folders.nodes, { folderName, parentFolderId });
+  await Folder.findByIdAndUpdate(folders._id, folders);
 
-  const existingFolder = await Folder.findOne({
-    name: folderName,
+  const parentFolder = searchByFolderId(folders.nodes, parentFolderId);
+  const folderVersion = new FolderVersion({
     workspace: workspaceId,
-    environment: environment,
-    parent: parentFolderId,
-    path: normalizedCurrentPath
+    environment,
+    // root condition
+    nodes: parentFolder || folders.nodes,
   });
+  await folderVersion.save();
+
+  return res.json({ folder });
+};
+
+export const updateFolderById = async (req: Request, res: Response) => {
+  const { folderId } = req.params;
+  const { name, workspaceId, environment } = req.body;
 
-  if (existingFolder) {
-    return res.json(existingFolder)
+  const folders = await Folder.findOne({ workspace: workspaceId, environment });
+  if (!folders) {
+    throw BadRequestError({ message: "The folder doesn't exist" });
   }
 
-  const newFolder = new Folder({
-    name: folderName,
+  // check that user is a member of the workspace
+  await validateMembership({
+    userId: req.user._id.toString(),
+    workspaceId,
+    acceptedRoles: [ADMIN, MEMBER],
+  });
+
+  const parentFolder = getParentFromFolderId(folders.nodes, folderId);
+  if (!parentFolder) {
+    throw BadRequestError({ message: "The folder doesn't exist" });
+  }
+  const folder = parentFolder.children.find(({ id }) => id === folderId);
+  if (!folder) {
+    throw BadRequestError({ message: "The folder doesn't exist" });
+  }
+
+  parentFolder.version += 1;
+  folder.name = name;
+  folder.version += 1;
+
+  await Folder.findByIdAndUpdate(folders._id, folders);
+  const folderVersion = new FolderVersion({
     workspace: workspaceId,
-    environment: environment,
-    parent: parentFolderId,
-    path: normalizedCurrentPath,
-    parentPath: normalizedParentPath
+    environment,
+    nodes: parentFolder,
   });
+  await folderVersion.save();
 
-  await newFolder.save();
+  // decision to take snapshot ot not
+  const secretCount = await Secret.count({
+    workspace: workspaceId,
+    environment,
+    folder: folderId,
+  });
+  if (secretCount || folder.children.length) {
+    await EESecretService.takeSecretSnapshot({
+      workspaceId,
+      environment,
+      folderId: parentFolder.name === "root" ? undefined : parentFolder.id,
```

Comment:
Is there a reason you have choose to denote root as `undefined` instead of the string root?

Reply:
Oopz my bad. Need to change back and just pass root